source/mir/random/flex/package.d

```diff
@@ -0,0 +1,1012 @@
+
+    // check last c
+    if (points[$ - 1].isInfinity)
+        assert(cs[$ - 1] > - 1,"cs must be > -1 for unbounded domains");
+}
+body
+{
+    import mir.random.flex.internal.calc: arcmean, calcInterval;
+    import mir.random.flex.internal.transformations : transform, transformInterval;
+    import mir.random.flex.internal.types: Interval;
+    import mir.internal.math: copysign, exp, pow;
+    import mir.sum: Summator, Summation;
+    import std.algorithm.sorting : sort;
+    import std.container.array: Array;
+    import std.container.binaryheap: BinaryHeap;
+    import std.math: nextDown;
+    import std.range.primitives : front, empty, popFront;
+
+    alias Sum = Summator!(S, Summation.precise);
+
+    Sum totalHatAreaSummator = 0;
+    Sum totalSqueezeAreaSummator = 0;
+
+    auto nrIntervals = cs.length;
+
+    // binary heap can't be extended with normal arrays, see
+    // https://github.com/dlang/phobos/pull/4359
+    auto arr = Array!(Interval!S)();
+    auto ips = BinaryHeap!(typeof(arr), (Interval!S a, Interval!S b){
+        S aVal = a.hatArea - a.squeezeArea;
+        S bVal = b.hatArea - b.squeezeArea;
+        // Explicit order is needed for LDC (undefined otherwise)
+        if (!(aVal == bVal))
+            return aVal < bVal;
+        else
+            return a.lx < b.lx;
+    })(arr);
+
+    S l = points[0];
+    S l0 = f0(points[0]);
+    S l1 = f1(points[0]);
+    S l2 = f2(points[0]);
+
+    version(Flex_logging)
+    {
+        log("starting flex with p=", points, ", cs=", cs, " rho=", rho);
+        version(Flex_logging_hex)
+        {
+            logf("points= %(%a, %)", points);
+            logf("cs= %(%a, %)", points);
+        }
+    }
+
+    // initialize with user given splitting points
+    foreach (i, r; points[1..$])
+    {
+        auto iv = Interval!S(l, r, cs[i], l0, l1, l2, f0(r), f1(r), f2(r));
+        l = r;
+        l0 = iv.rtx;
+        l1 = iv.rt1x;
+        l2 = iv.rt2x;
+        transformInterval(iv);
+
+        calcInterval(iv);
+        totalHatAreaSummator += iv.hatArea;
+        totalSqueezeAreaSummator += iv.squeezeArea;
+
+        ips.insert(iv);
+    }
+
+    version(Windows) {} else
+    version(Flex_logging)
+    {
+        import std.algorithm.iteration : map;
+        import std.array : array;
+        auto ipsD = ips.dup.array;
+        log("----");
+
+        logf("Interval: %(%f, %)", ipsD.map!`a.lx`);
+        version(Flex_logging_hex)
+            logf("Interval: %(%a, %)", ipsD.map!`a.lx`);
+
+        logf("hatArea: %(%f, %)", ipsD.map!`a.hatArea`);
+        version(Flex_logging_hex)
+            logf("hatArea: %(%a, %)", ipsD.map!`a.hatArea`);
+
+        logf("squeezeArea %(%f, %)", ipsD.map!`a.squeezeArea`);
+        version(Flex_logging_hex)
+            logf("squeezeArea %(%a, %)", ipsD.map!`a.squeezeArea`);
+
+        log("----");
+    }
+
+    // Flex is not guaranteed to converge
+    for (auto i = 0; i < maxIterations && nrIntervals < apprMaxPoints; i++)
```

Comment:
If the code is correct `maxIterations` = `approxMaxPoints - startingPoints.length`. I think it's more intuitive for the user to specify only `approxMaxPoints`. @9il if you don't object I will remove `maxIterations`.